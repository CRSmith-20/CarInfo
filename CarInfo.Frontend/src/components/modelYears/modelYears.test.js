import {shallow, mount} from 'enzyme';
import ModelYears from './modelYears';
import axios from "axios";
import { BrowserRouter as Router } from 'react-router-dom';
import { act } from "react-dom/test-utils";
import toJson from 'enzyme-to-json';
import { createMemoryHistory } from 'history'

jest.mock("axios");

let formattedPairs;
let match;
let wrapper;
let history;

beforeEach(() => {
    formattedPairs = [
        {ID: 1, Year: 1990}, {ID: 2, Year: 2220}, {ID: 3, Year: 3330}
    ]
    match = { params: { model: 'Test' } }

    history = createMemoryHistory();
    history.location.key = 'consistent'; //needed for snapshot as it's unique per run

});

describe('<ModelYears />', () => {
    describe('when component initalizes', () => {
        it('should initalize to loading', () => {
            const test = shallow(<Router><ModelYears /></Router>);
            expect(toJson(test)).toMatchSnapshot();
        });
    });
    
    describe('when component loads', () => {
        it('should render makes with links', async () => {
            
            
            await act(async () => {
                axios.get.mockResolvedValueOnce({data: formattedPairs});
                wrapper = mount(
                    <Router>
                        <ModelYears history={history} match={match} />
                    </Router>
                );
            });
    
            wrapper.update();

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('Link').length).toBe(3);
        })
    });
    
    describe('when api calls fail', () => {
        it('should render a ErrorDisplay', async () => {                      
            await act(async () => {
                axios.get.mockRejectedValueOnce(new Error("Generic Failure"));
                wrapper = mount(<Router><ModelYears history={history} match={match} /></Router>);
            });
    
            wrapper.update();

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('ErrorDisplay').length).toBe(1);
        })
    });
})