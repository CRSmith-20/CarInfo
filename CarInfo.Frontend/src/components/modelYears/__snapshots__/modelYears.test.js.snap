// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModelYears /> when component initalizes should initalize to loading 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ModelYears />
</Router>
`;

exports[`<ModelYears /> when component loads should render years with links 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ModelYears
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "consistent",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "consistent",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      match={
        Object {
          "params": Object {
            "model": "Test",
          },
        }
      }
    >
      <div>
        <div>
          <div
            key="1"
          >
            <Link
              to="/details/Test/1990/1"
            >
              <LinkAnchor
                href="/details/Test/1990/1"
                navigate={[Function]}
              >
                <a
                  href="/details/Test/1990/1"
                  onClick={[Function]}
                >
                  1990
                </a>
              </LinkAnchor>
            </Link>
          </div>
          <div
            key="2"
          >
            <Link
              to="/details/Test/2220/2"
            >
              <LinkAnchor
                href="/details/Test/2220/2"
                navigate={[Function]}
              >
                <a
                  href="/details/Test/2220/2"
                  onClick={[Function]}
                >
                  2220
                </a>
              </LinkAnchor>
            </Link>
          </div>
          <div
            key="3"
          >
            <Link
              to="/details/Test/3330/3"
            >
              <LinkAnchor
                href="/details/Test/3330/3"
                navigate={[Function]}
              >
                <a
                  href="/details/Test/3330/3"
                  onClick={[Function]}
                >
                  3330
                </a>
              </LinkAnchor>
            </Link>
          </div>
        </div>
        <button
          onClick={[Function]}
        >
          Return to Models
        </button>
      </div>
    </ModelYears>
  </Router>
</BrowserRouter>
`;

exports[`<ModelYears /> when the api call fails should render a ErrorDisplay 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ModelYears
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "consistent",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "consistent",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      match={
        Object {
          "params": Object {
            "model": "Test",
          },
        }
      }
    >
      <ErrorDisplay
        title="Select Year"
      >
        <div>
          <h1
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            Select Year
          </h1>
          <Alert
            closeLabel="Close alert"
            key="error"
            show={true}
            style={
              Object {
                "textAlign": "center",
              }
            }
            transition={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "appear": false,
                  "in": false,
                  "mountOnEnter": false,
                  "timeout": 300,
                  "unmountOnExit": false,
                },
                "render": [Function],
              }
            }
            variant="warning"
          >
            <Fade
              appear={false}
              in={true}
              mountOnEnter={false}
              style={
                Object {
                  "textAlign": "center",
                }
              }
              timeout={300}
              unmountOnExit={true}
            >
              <Transition
                addEndListener={[Function]}
                appear={false}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                style={
                  Object {
                    "textAlign": "center",
                  }
                }
                timeout={300}
                unmountOnExit={true}
              >
                <div
                  className="fade alert alert-warning show"
                  role="alert"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  An error has occurred while loading, please refresh and try again.
                </div>
              </Transition>
            </Fade>
          </Alert>
        </div>
      </ErrorDisplay>
    </ModelYears>
  </Router>
</BrowserRouter>
`;
